pkg/interface/src/apps/chat/components/lib/chat-editor.js

```diff
@@ -0,0 +1,134 @@
+import React, { Component } from 'react';
+import { UnControlled as CodeEditor } from 'react-codemirror2';
+import CodeMirror from 'codemirror';
+
+import 'codemirror/mode/markdown/markdown';
+import 'codemirror/addon/display/placeholder';
+
+import 'codemirror/lib/codemirror.css';
+
+const BROWSER_REGEX =
+  new RegExp(String(!/Android|webOS|iPhone|iPad|iPod|BlackBerry/i));
+
+
+const MARKDOWN_CONFIG = {
+  name: 'markdown',
+  tokenTypeOverrides: {
+    header: 'presentation',
+    quote: 'presentation',
+    list1: 'presentation',
+    list2: 'presentation',
+    list3: 'presentation',
+    hr: 'presentation',
+    image: 'presentation',
+    imageAltText: 'presentation',
+    imageMarker: 'presentation',
+    formatting: 'presentation',
+    linkInline: 'presentation',
+    linkEmail: 'presentation',
+    linkText: 'presentation',
+    linkHref: 'presentation'
+  }
+};
+
+export default class ChatEditor extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      message: props.message
+    };
+    this.editor = null;
+  }
+
+  componentWillUnmount() {
+    this.props.onUnmount(this.state.message);
+  }
+
+  componentDidUpdate(prevProps) {
+    const { props } = this;
+    if (prevProps.message !== props.message) {
+      this.editor.setValue(props.message);
+      this.editor.setOption('mode', MARKDOWN_CONFIG);
+      return;
+    }
+
+    if (!props.inCodeMode) {
+      this.editor.setOption('mode', MARKDOWN_CONFIG);
+      this.editor.setOption('placeholder', this.props.placeholder);
+    } else {
+      this.editor.setOption('mode', null);
+      this.editor.setOption('placeholder', 'Code...');
+    }
+    const value = this.editor.getValue();
+
+    // Force redraw of placeholder
+    if(value.length === 0) {
+      this.editor.setValue(' ');
+      this.editor.setValue('');
+    }
+  }
+
+  submit() {
+    if(!this.editor) {
+      return;
+    }
+
+    let editorMessage = this.editor.getValue();
+    if (editorMessage === '') {
+      return;
+    }
+
+    this.props.submit(editorMessage);
+    this.editor.setValue('');
+  }
+
+  messageChange(editor, data, value) {
+    if (value == this.props.message || value == '' || value == ' ') {
+      return;
+    }
+    this.setState({
+      message: value
+    });
+  }
+
+  render() {
+    const { props } = this;
+
+    const codeTheme = props.inCodeMode ? ' code' : '';
+
+    const options = {
+      mode: MARKDOWN_CONFIG,
+      theme: 'tlon' + codeTheme,
+      lineNumbers: false,
+      lineWrapping: true,
+      scrollbarStyle: 'native',
+      cursorHeight: 0.85,
+      placeholder: props.inCodeMode ? 'Code...' : props.placeholder,
+      extraKeys: {
+        'Enter': () => {
+          this.submit();
+        }
```

Comment:
We have more keys defined here — `Shift-3` to toggle code mode for example (and only if the message was empty). 

Though honestly, I'm unsure if anyone used this shortcut — it was meant to call back to how code used to be prefixed with `#`, and it's not documented for users.

@urcades

Reply:
Are you thinking we remove this? It's a strange input expectation, there's some crossed wires here re: how we treat markdown in general in this input, the code entry mechanism, etc.

Reply:
I think it would be fine to just remove it because it's indeed a strange expectation and in general I'd like to make UX less, well, unintuitive. But I was in any case just flagging that it was removed so we could assent to that

Reply:
Yeah — I'm in favor of removing

Reply:
Ok gonna leave this as is